presto-hive/src/main/java/parquet/io/ColumnIOConverter.java

```diff
@@ -0,0 +1,97 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package parquet.io;
+
+import com.facebook.presto.hive.parquet.Field;
+import com.facebook.presto.hive.parquet.GroupField;
+import com.facebook.presto.hive.parquet.PrimitiveField;
+import com.facebook.presto.hive.parquet.RichColumnDescriptor;
+import com.facebook.presto.spi.type.MapType;
+import com.facebook.presto.spi.type.NamedTypeSignature;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeSignatureParameter;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Locale;
+import java.util.Optional;
+
+import static com.facebook.presto.hive.parquet.ParquetTypeUtils.getArrayElementColumn;
+import static com.facebook.presto.hive.parquet.ParquetTypeUtils.getMapKeyValueColumn;
+import static com.facebook.presto.spi.type.StandardTypes.ARRAY;
+import static com.facebook.presto.spi.type.StandardTypes.MAP;
+import static com.facebook.presto.spi.type.StandardTypes.ROW;
+
+/**
+ * Placed in parquet.io package to have access to ColumnIO getRepetitionLevel() and getDefinitionLevel() methods.
+ */
+public class ColumnIOConverter
+{
+    private ColumnIOConverter()
+    {
+    }
+
+    public static Optional<Field> constructField(Type type, ColumnIO columnIO)
+    {
+        if (columnIO == null) {
+            return Optional.empty();
+        }
+        boolean required = columnIO.getType().getRepetition() != parquet.schema.Type.Repetition.OPTIONAL;
+        int repetitionLevel = columnIO.getRepetitionLevel();
+        int definitionLevel = columnIO.getDefinitionLevel();
+        if (ROW.equals(type.getTypeSignature().getBase())) {
+            GroupColumnIO groupColumnIO = (GroupColumnIO) columnIO;
+            List<Type> parameters = type.getTypeParameters();
+            ImmutableList.Builder<Optional<Field>> fileldsBuilder = ImmutableList.builder();
+            List<TypeSignatureParameter> fields = type.getTypeSignature().getParameters();
+            boolean structHasParameters = false;
+            for (int i = 0; i < fields.size(); i++) {
+                NamedTypeSignature namedTypeSignature = fields.get(i).getNamedTypeSignature();
+                String name = namedTypeSignature.getName().get().toLowerCase(Locale.ENGLISH);
+                Optional<Field> field = constructField(parameters.get(i), groupColumnIO.getChild(name));
+                structHasParameters |= field.isPresent();
+                fileldsBuilder.add(field);
+            }
+            if (structHasParameters) {
+                return Optional.of(new GroupField(type, repetitionLevel, definitionLevel, required, fileldsBuilder.build()));
+            }
+            else {
```

Comment:
this `else` is unnecessary, we can just return.

Reply:
done